### ISSUE TITLE
"(Tcl) invalid command name" when setting iemgui send/receive name in a vis 0 canvas

### ISSUE THREAD

**Issue by claudeha** [pure-data/pure-data#477]
to reproduce, click the message box:

    #N canvas 40 73 450 300 10;
    #N canvas 760 223 450 300 subpatch 0;
    #X obj 212 130 tgl 15 0 empty r r 17 7 0 10 -262144 -1 -1 0 1;
    #X restore 15 123 pd subpatch;
    #X msg 12 59 \; pd-subpatch vis 0 \; r receive empty \, receive r;

resulting error:

    (Tcl) INVALID COMMAND NAME: invalid command name ".x55895a47e490.c"
        while executing
    ".x55895a47e490.c create rectangle 212 130 219 131 -fill black -tags 55895a47e7a0IN0"
        ("uplevel" body line 1)
        invoked from within
    "uplevel #0 $docmds"

root cause:

pd/pd-gui communication is fragile, checking that a bunch of stuff from the socket ends in "\n" and is "complete" tcl.  If an error is thrown at the start of the bunch of stuff, a non-deterministic amount of subsequent stuff may be skipped or lost, depending on network conditions and so on.

for example, opening the subpatch and clicking the message box (with pd-gui patched to print the Tcl source code on error) , I get:

    (Tcl) INVALID COMMAND NAME: 
    invalid command name ".x55dd55e157c0.c"
        while executing
    ".x55dd55e157c0.c create rectangle 212 130 219 131 -fill black -tags 55dd55e15ad0IN0"
        ("uplevel" body line 4)
        invoked from within
    "uplevel #0 $docmds"
    (Tcl) INVALID COMMAND NAME: 
    invalid command name ".x55dd55e157c0.c"
        while executing
    ".x55dd55e157c0.c create rectangle 212 130 219 131 -fill black -tags 55dd55e15ad0IN0"
        ("uplevel" body line 4)
        invoked from within
    "uplevel #0 $docmds"
    (Tcl) COMMAND SOURCE CODE: 
    .x55dd55e157c0.c delete all
    destroy .x55dd55e157c0
    ::pd_menus::update_window_menu
    .x55dd55e157c0.c create rectangle 212 130 219 131 -fill black -tags 55dd55e15ad0IN0
    .x55dd55e157c0.c raise 55dd55e15ad0LABEL 55dd55e15ad0IN0
    .x55dd55e157c0.c delete 55dd55e15ad0IN0
    .x55dd55e15160.c itemconfigure .x55dd55e15160.t55dd55e16390R -width 5
    
where you can see that the first 3 commands completed succesfully, the 4th failed, and the remainder was not executed.

**Comment by Spacechild1** [comment]
I can confirm this for Pd 0.48 and 0.49. Your observations about Tcl error handling aside, I think the root problem here is rather that for some reason you can't set the receive name if the subpatch is not mapped while other actions, like changing the color, *do* work.

**Comment by umlaeute** [comment]
it's most likely a problem with disabling the inlet/outlet rectangle.

**Comment by millerpuckette** [comment]
reopening this.... all of the fixes weve tried seem to cause even bigger problems.

**Comment by katjav** [comment]
This can't be solved in g_allguis.c because each iemgui has its own method to draw / erase IOlets, which is called when names are changed from send/receivable to non-send/receivable, currently regardless of vis state. Probably the vis state should be checked in these methods, or was that already tried in an earlier commit ? (I lost track of them).

**Comment by katjav** [comment]
Tried the above idea and can confirm that this resolves the issue at least for the example patch from @claudeha. Also it is still possible to erase and redraw IOlets when the canvas is visible. So this seems a simple fix (but I don't know about relations with other issues). Just to be clear, what I did was adding the checks in iemgui_send() and iemgui_receive() such that their last lines look like:

```
if(glist_isvisible(iemgui->x_glist)) 
    (*iemgui->x_draw)(x, iemgui->x_glist, IEM_GUI_DRAW_MODE_IO + oldsndrcvable);
```
